askama_escape/src/lib.rs

```diff
@@ -50,45 +53,212 @@ pub fn escape(s: &str) -> Escaped {
+            _ => (),
+        }
+    }};
+}
+
+macro_rules! write_char {
+    ($i:expr, $b: expr, $start: ident, $fmt: ident, $bytes:ident) => {{
+        let b = $b;
+        if b.wrapping_sub(FLAG_BELOW) <= LEN {
+            escaping_bodies!($i, b, $start, $fmt, $bytes);
+        }
+    }};
+}
+
+macro_rules! write_mask {
+    ($cmp: ident, $ptr: ident, $start_ptr: ident, $start: ident, $fmt: ident, $bytes:ident) => {{
+        let mut mask = _mm256_movemask_epi8($cmp);
+        if mask != 0 {
+            let at = sub($ptr, $start_ptr);
+            let mut cur = mask.trailing_zeros() as usize;
+
+            loop {
+                let b = *$ptr.add(cur);
+                escaping_bodies!(at + cur, b, $start, $fmt, $bytes);
+
+                mask ^= 1 << cur;
+                if mask == 0 {
+                    break;
+                }
+
+                cur = mask.trailing_zeros() as usize;
+            }
+        }
+    }};
+}
+
 pub struct Escaped<'a> {
     bytes: &'a [u8],
 }
 
-impl<'a> ::std::fmt::Display for Escaped<'a> {
+impl<'a> fmt::Display for Escaped<'a> {
     fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
-        let mut start = 0;
-        for (i, b) in self.bytes.iter().enumerate() {
-            if b.wrapping_sub(b'"') <= FLAG {
-                match *b {
-                    b'<' => escaping_body!(start, i, fmt, self, "&lt;"),
-                    b'>' => escaping_body!(start, i, fmt, self, "&gt;"),
-                    b'&' => escaping_body!(start, i, fmt, self, "&amp;"),
-                    b'"' => escaping_body!(start, i, fmt, self, "&quot;"),
-                    b'\'' => escaping_body!(start, i, fmt, self, "&#x27;"),
-                    b'/' => escaping_body!(start, i, fmt, self, "&#x2f;"),
-                    _ => (),
-                }
+        _imp(self.bytes, fmt)
+    }
+}
+
+cfg_if! {
+    if #[cfg(all(target_arch = "x86_64"))] {
+
+        use std::arch::x86_64::*;
+        use std::mem::size_of;
+        const VECTOR_SIZE: usize = size_of::<__m256i>();
+        const VECTOR_ALIGN: usize = VECTOR_SIZE - 1;
+        const LOOP_SIZE: usize = 2 * VECTOR_SIZE;
+
+        // Subtract `b` from `a` and return the difference. `a` should be greater than
+        // or equal to `b`.
+        #[inline(always)]
+        fn sub(a: *const u8, b: *const u8) -> usize {
+            debug_assert!(b <= a);
+            (a as usize) - (b as usize)
+        }
+
+        #[inline(always)]
+        fn _imp(bytes: &[u8], fmt: &mut fmt::Formatter) -> fmt::Result {
+            unsafe { _avx_escape(bytes, fmt) }
+        }
+    } else {
+
+        #[inline(always)]
+        fn _imp(bytes: &[u8], fmt: &mut fmt::Formatter) -> fmt::Result {
+            _escape(bytes, fmt)
+        }
+    }
+}
+
+#[inline(always)]
+fn _escape(bytes: &[u8], fmt: &mut fmt::Formatter) -> fmt::Result {
+    let mut start = 0;
+
+    for (i, b) in bytes.iter().enumerate() {
+        unsafe {
```

Comment:
Why do we have the large `unsafe` block here, instead of the smaller one before?

Reply:
Ok, or directly `unsafe fn`?

Reply:
I don't understand your question. My concern is that the new `unsafe {}` block covers a lot of code, so it's not immediately obvious which parts are actually unsafe.

Reply:
is by `str::from_unchecked_utf-8` in `write_char!
`

Reply:
`write_char` is used in `_avx_escape` which is `unsafe` and returns the double unsafe if I just decorate this instruction `str::from_unch..`
.

Reply:
I have marked all the unsafe blocks of the macros with comments. This block picks up a lot so I do not repeat myself.

Reply:
of the unsafe topic, can you send a pull request to my master?